### PR TITLE
[Dev] FP8 params support for megatron-fsdp (MXFP8/Blockwise)

### PR DESCRIPTION
**Pull request by kunlunl**
# What does this PR do ?

main MR: https://github.com/NVIDIA/Megatron-LM/pull/2239

(Merged) TE MR: https://github.com/NVIDIA/TransformerEngine/pull/2055

Major changes:
* Make Megatron-FSDP’s AG pipeline support using different data parallel buffers, because MXFP8 has different quantization direction in forward and backward passes.
* Decouple the FP8-related logic from the main workflow and provide a unified abstraction to 1) operate the raw data storage of different recipes; 2) Create or discard transpose cache for different recipes.

## Convergence Test

Comparison of convergence behavior across several configurations:

- Distopt-BF16
- Distopt-CSFP8
- MFSDP-CSFP8
- MFSDP-MXFP8
- MFSDP-BlockFP8

Screenshot 2025-12-17 at 13 57 28

https://api.wandb.ai/links/nvidia/sga6o3t3

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Review by shjwudp — CHANGES_REQUESTED** [review]
Please resolve the conflicts and make sure all tests pass.

I think naming the file that handles mixed precision as _mixed_precision.py_ would be more appropriate.

**Comment by shjwudp** [comment]
Tested on Llama3-8B. Convergence was stable across all configurations.

Performance Result: 

- M-FSDP-mxfp8 achieved a 34% speedup over M-FSDP-bf16 (1385 TFLOPs vs. 1033 TFLOPs).
- M-FSDP-mxfp8 reduced memory usage by 21.93 GB (155.89 GB vs. 177.82 GB for bf16).

Full test report: https://api.wandb.ai/links/nvidia/z3nax4om

**Comment by shjwudp** [comment]
> * CICD Megatron-LM / tests/unit_tests/distributed/fsdp/**/*.py - latest (push)

@yanring Sure, we will follow up.

**Comment by yanring** [comment]
@kunlunl Could you help follow up on the merging? This has been approved for a while.

**Comment by kunlunl** [comment]
> @kunlunl Could you help follow up on the merging? This has been approved for a while.

@yanring  We are waiting for the merging of TE PR https://github.com/NVIDIA/TransformerEngine/pull/2055, in case there's any change in the TE API.

**Comment by yanring** [comment]
@kunlunl @shjwudp pls let me know whenever we want to proceed

**Comment by yanring** [comment]
# Code Review: PR #2086 - [Dev] FP8 params support for megatron-fsdp

## 1. PR Summary

**What this PR does:**
This PR adds FP8 parameter support for Megatron-FSDP, specifically enabling MXFP8 (Microscaling FP8) which requires different quantization directions in forward and backward passes. The key architectural change is supporting separate `transpose_weight_buffer` for MXFP8 parameters alongside the existing `model_weight_buffer`.

**Why it's needed:**
MXFP8 has a unique characteristic where forward and backward passes require different data representations (row-wise vs column-wise storage). The existing FSDP implementation assumed a single weight buffer per parameter group, which doesn't accommodate this requirement.

**How it's implemented:**
1. **New abstraction layer (`mixed_precision.py`)**: Decouples FP8-related logic into a unified module providing operations on FP8 tensors across different TE versions
2. **Dual buffer architecture**: Adds `transpose_weight_buffer` to `ParameterGroup` for MXFP8 parameters
3. **Bucket key modification**: Changes bucket tracking from `bucket_id` to `(bucket_id, bwd)` tuple to manage forward/backward buffers independently
4. **All-gather pipeline updates**: Modified to handle separate buffers for forward and backward passes via `bwd` parameter

---

## 2. Critical Issues [P0/P1]

### **[P0]** `param_and_grad_buffer.py:2734` - Logic Error in `reduce_scatter_gradients`

```python
for g in self.parameter_groups:
    gbuf = g.main_grad_buffer
    if gbuf is not None:
        continue  # TODO(mxfp8): This is an error?
```

- **Problem**: The condition `if gbuf is not None: continue` skips processing when the gradient buffer **exists**. This means gradient reduce-scatter will only happen when there's NO gradient buffer, which is the opposite of intended behavior. The TODO comment even acknowledges this might be an error.
- **Fix**: The condition should be `if gbuf is None: continue` to skip parameter groups without gradient buffers.

```python
# Suggested fix
if gbuf is None:
    continue
```

---

### **[P1]** `mixed_precision.py:362` - Type Signature Mismatch with Implementation

```python
def fp8_create_transpose_cache(tensors: List[torch.Tensor]) -> None:
    """Create the transpose cache of a FP8 tensor."""
```

- **Problem**: The docstring says "a FP8 tensor" (singular), but the parameter is `List[torch.Tensor]`. More critically, the caller at `megatron_fsdp.py:471` passes a single parameter: `fp8_create_transpose_cache(param)`, not a list.
- **Fix**: Either change the signature to accept `Union[torch.Tensor, List[torch.Tensor]]` or update the caller to pass `[param]`.

```python
# Suggested fix - update the signature and add input normalization
def fp8_create_transpose_cache(tensors: Union[torch.Tensor, List[torch.Tensor]]) -> None:
    """Create the transpose cache of FP8 tensor(s)."""
    if not isinstance(tensors, list):
        tensors = [tensors]
    # ... rest of implementation
```

---

### **[P1]** `param_and_grad_buffer.py:3166-3170` - Incorrect HSDP Branch in `get_fsdp_buffer`

```python
def get_fsdp_buffer(self, bucket_id: int, bwd=False) -> DataParallelBuffer:
    """Get the FSDP buffer with the given bucket ID."""
    param_group = self.buffer.parameter_groups[bucket_id]
    if self.buffer.ddp_config.outer_dp_sharding_strategy != "no_shard":
        if bwd and param_group.transpose_weight_buffer is not None:
            raise RuntimeError("Transpose buffer is not supported for HSDP")
        else:
            return param_group.model_weight_buffer  # Bug: should return hsdp_wbuf
```

- **Problem**: When using HSDP (outer_dp_sharding_strategy != "no_shard"), the function returns `model_weight_buffer` instead of `hsdp_wbuf`. This breaks HSDP functionality. The original code had `return param_group.hsdp_wbuf`.
- **Fix**: Restore the correct HSDP buffer return logic.

```python
# Suggested fix
if self.buffer.ddp_config.outer_dp_sharding_strategy != "no_shard":
    if bwd and param_group.transpose_weight_buffer is not None:
        raise RuntimeError("Transpose buffer is not supported for HSDP")
    return param_group.hsdp_wbuf  # Correct: return HSDP buffer
if bwd and param_group.transpose_weight_buffer is not None:
    return param_group.transpose_weight_buffer
return param_group.model_weight_buffer
```

---

### **[P1]** `megatron_fsdp.py:611-614` - Unnecessary Double All-Gather

```python
# TODO(mxfp8): Do we really need this?
self.all_gather_and_wait_parameters_ready(
    params=param_list, prefetch=False, bwd=True
)
```

- **Problem**: The TODO comment questions necessity, but this adds a full all-gather operation during pre-forward for **all** `bwd=True` cases, not just MXFP8. This could cause significant performance regression for non-MXFP8 FP8 modes and even non-FP8 training.
- **Fix**: Gate this behind a check for MXFP8 parameters, or investigate if it's truly needed.

```python
# Suggested fix - only do this for MXFP8 parameters
if any(fp8_need_transpose_data(p) for p in param_list if is_float8tensor(p)):
    self.all_gather_and_wait_parameters_ready(
        params=param_list, prefetch=False, bwd=True
    )
```

---

### **[P1]** Missing Unit Tests for MXFP8 FSDP

- **Problem**: The existing test file `test_mfsdp_fully_shard.py` does not include test cases for FP8 parameters or MXFP8 specifically. This is a significant feature addition without corresponding test coverage.
- **Fix**: Add test cases that:
  1. Test FSDP with FP8 parameters (`--fp8-param-gather`)
  2. Test FSDP with MXFP8 recipe specifically
  3. Verify that transpose buffers are correctly managed

---

## 3. Suggestions [P2]

### **[P2]** `mixed_precision.py:230` - Bare `except:` Clause

```python
except:
    TE_VERSION = None
```

- **Suggestion**: Use explicit exception types instead of bare `except:`. This could silently catch keyboard interrupts or system exit.

```python
except (ImportError, AttributeError, Exception) as e:
    logger.debug(f"Could not determine TE version: {e}")
    TE_VERSION = None
```

---

### **[P2]** `mixed_precision.py:239-246` - Multiple Bare `except:` Clauses

Similar bare `except:` clauses appear for detecting FP8 tensor class and MXFP8 tensor class. Apply the same fix.

---

### **[P2]** `param_and_grad_buffer.py:2982-2984` - NotImplementedError for HSDP Transpose Buffer

```python
if group.transpose_weight_buffer is not None:
    raise NotImplementedError("HSDP for transpose buffer is not implemented yet")
```

- **Suggestion**: This raises at runtime during initialization. Consider adding an early validation check in the configuration phase with a clearer error message.

---

### **[P2]** `param_and_grad_buffer.py:3628-3644` - Simplified Error Handling in Override Methods

```python
def override_sharded_param_to_function(*args, **kwargs):
    if p._typed_storage()._size() == 0:
        warnings.warn(
            "The parameter may be sharded by Megatron-FSDP, "
            "no actual 'to' operation is performed."
        )
        return torch.empty([])
    return to_function(*args, **kwargs)
```

- **Suggestion**: The original code checked `bucket_status` which was more precise. The new approach uses storage size as a proxy, which may not always be accurate. Consider if this simplification could cause issues.

---

### **[P2]** `megatron_fsdp.py:1002-1011` - Multiple TODOs Without Tracking

Multiple `# TODO(mxfp8):` comments without clear tracking:
- Line 611: "Do we really need this?"
- Line 1002: "Is this correct?"
- Line 1013: "Is this correct?"
- Line 1023: "Is this correct?"

- **Suggestion**: These TODOs indicate uncertainty in the implementation. Consider:
  1. Creating GitHub issues to track these
  2. Adding more detailed comments explaining the reasoning
  3. Adding assertions or logging to validate correctness at runtime

---

### **[P2]** `param_and_grad_buffer.py:1801-1802` - TODO Comment About Allocator

```python
# TODO(mxfp8): Do we need separate alloc for transpose buffer?
```

- **Suggestion**: This is an important architectural question. Sharing allocators could cause memory fragmentation or contention. Document the decision or create a follow-up issue.

**Comment by kunlunl** [comment]
@yanring It's ready to be proceed.

**Comment by cspades** [comment]
@kunlunl @shjwudp Can we also add a simple non-MLM unit test for FP8 using `fully_shard`?

I can work on this as well and get it in a subsequent PR. Just curious if there's any blockers / should be as easy as `fp8_model_init` + `fp8_autocast`.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 1]
```diff
@@ -558,12 +565,15 @@ def _post_backward(module, *unused):
...
                     # because we have our data-parallel gradients computed.
-                    release_module_parameters(module)
+                    release_module_parameters(module, bwd=True)
                     module._training_state = TrainingState.IDLE
                 param_list = list(module.parameters())
             else:
                 param_list = list(module.parameters(recurse=False))
 
+            if self.enable_fine_grained_param_gather_hook:
+                param_list = list(module.parameters(recurse=False))
```
Comment (cspades) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:575]: @kunlunl Basic question, why does MXFP8 need fine-grained AG again?

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:575]: Good question. This logic was added to support MXFP8 + activation recompute:

1. It ensures the accuracy of the all-gather sub-module parameters — the parameters used in the backward all-gather cannot be reused in the forward pass in MXFP8.
2. It improves the performance of gradient accumulation and gradient reduction. Fine-grained handling allows earlier launches and prevents potential hangs during MoE training.

There seems to be an issue with point (1), and there may be alternative solutions to address it. : /

Comment (kunlunl) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:575]: Yes, for "the parameters used in the backward all-gather cannot be reused in the forward pass in MXFP8", it's because mxfp8 is 1D scaling, so its row-wise data and column-wise have different values, we cannot get the column-wise data by transposing the row-wise data.
